File: golang/sfv_test.go
package httpsig

import "testing"

func TestSerializeSignatureParams(t *testing.T) {
	tests := []struct {
		name       string
		components []ComponentIdentifier
		params     *SFVParams
		want       string
	}{
		{
			name:       "empty components, no params",
			components: nil,
			params:     nil,
			want:       "()",
		},
		{
			name:       "empty components with created and keyid",
			components: []ComponentIdentifier{},
			params: func() *SFVParams {
				p := NewSFVParams()
				p.Set("created", int64(1618884473))
				p.Set("keyid", "test-key")
				return p
			}(),
			want: `();created=1618884473;keyid="test-key"`,
		},
		{
			name: "multiple components",
			components: []ComponentIdentifier{
				Component("@method"),
				Component("@path"),
				Component("content-type"),
			},
			params: func() *SFVParams {
				p := NewSFVParams()
				p.Set("created", int64(1618884473))
				p.Set("keyid", "my-key")
				return p
			}(),
			want: `("@method" "@path" "content-type");created=1618884473;keyid="my-key"`,
		},
		{
			name: "component with params",
			components: []ComponentIdentifier{
				Component("@authority"),
				QueryParam("Pet"),
			},
			params: func() *SFVParams {
				p := NewSFVParams()
				p.Set("created", int64(1618884473))
				return p
			}(),
			want: `("@authority" "@query-param";name="Pet");created=1618884473`,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := SerializeSignatureParams(tt.components, tt.params)
			if got != tt.want {
				t.Errorf("got:  %s\nwant: %s", got, tt.want)
			}
		})
	}
}

func TestParseDictionary(t *testing.T) {
	tests := []struct {
		name    string
		input   string
		wantLen int
		wantErr bool
	}{
		{
			name:    "single binary item",
			input:   `sig1=:dGVzdA==:`,
			wantLen: 1,
		},
		{
			name:    "two items",
			input:   `sig1=:dGVzdA==:, sig2=:YW5vdGhlcg==:`,
			wantLen: 2,
		},
		{
			name:    "inner list",
			input:   `sig1=("@method" "@path");created=1234`,
			wantLen: 1,
		},
		{
			name:    "empty",
			input:   "",
			wantLen: 0,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			members, err := ParseDictionary(tt.input)
			if (err != nil) != tt.wantErr {
				t.Fatalf("err = %v, wantErr = %v", err, tt.wantErr)
			}
			if len(members) != tt.wantLen {
				t.Errorf("got %d members, want %d", len(members), tt.wantLen)
			}
		})
	}
}

func TestParseDictionaryInnerList(t *testing.T) {
	input := `sig1=("date" "@method" "@path");created=1618884473;keyid="test-key"`
	members, err := ParseDictionary(input)
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if len(members) != 1 {
		t.Fatalf("expected 1 member, got %d", len(members))
	}
	m := members[0]
	if m.Key != "sig1" {
		t.Errorf("key = %q, want %q", m.Key, "sig1")
	}
	if m.InnerList == nil {
		t.Fatal("expected inner list")
	}
	if len(m.InnerList.Items) != 3 {
		t.Fatalf("expected 3 items, got %d", len(m.InnerList.Items))
	}
	if m.InnerList.Items[0].Value != "date" {
		t.Errorf("item[0] = %v, want %q", m.InnerList.Items[0].Value, "date")
	}
	if m.InnerList.Params == nil {
		t.Fatal("expected params")
	}
	if created, ok := m.InnerList.Params.Get("created"); !ok || created != int64(1618884473) {
		t.Errorf("created = %v, want 1618884473", created)
	}
	if keyid, ok := m.InnerList.Params.Get("keyid"); !ok || keyid != "test-key" {
		t.Errorf("keyid = %v, want %q", keyid, "test-key")
	}
}

func TestParseDictionaryBinaryValue(t *testing.T) {
	input := `sig1=:dGVzdA==:`
	members, err := ParseDictionary(input)
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	if len(members) != 1 {
		t.Fatalf("expected 1 member, got %d", len(members))
	}
	m := members[0]
	if m.Item == nil {
		t.Fatal("expected item")
	}
	bs, ok := m.Item.Value.([]byte)
	if !ok {
		t.Fatalf("expected []byte, got %T", m.Item.Value)
	}
	if string(bs) != "test" {
		t.Errorf("value = %q, want %q", string(bs), "test")
	}
}

func TestSerializeBareItem(t *testing.T) {
	tests := []struct {
		name string
		val  any
		want string
	}{
		{"string", "hello", `"hello"`},
		{"int64", int64(42), "42"},
		{"bool true", true, "?1"},
		{"bool false", false, "?0"},
		{"binary", []byte("test"), ":dGVzdA==:"},
		{"token", Token("rsa-pss-sha512"), "rsa-pss-sha512"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := serializeBareItem(tt.val)
			if got != tt.want {
				t.Errorf("got %q, want %q", got, tt.want)
			}
		})
	}
}

File: typescript/test/sfv.test.ts
import { describe, it, expect } from "vitest";
import {
  newSFVParams,
  sfvParamsSet,
  serializeSignatureParams,
  serializeBareItem,
  parseDictionary,
  component,
  queryParam,
  componentWithParams,
  SFVToken,
} from "../src/index.js";

describe("SFV serialization", () => {
  it("serializes empty signature params", () => {
    const result = serializeSignatureParams([], newSFVParams());
    expect(result).toBe("()");
  });

  it("serializes signature params with components and metadata", () => {
    let params = newSFVParams();
    params = sfvParamsSet(params, "created", 1618884473);
    params = sfvParamsSet(params, "keyid", "test-key");

    const result = serializeSignatureParams(
      [component("@method"), component("@path")],
      params,
    );
    expect(result).toBe(
      '("@method" "@path");created=1618884473;keyid="test-key"',
    );
  });

  it("serializes parameterized components", () => {
    let params = newSFVParams();
    params = sfvParamsSet(params, "created", 1618884473);
    params = sfvParamsSet(params, "keyid", "test-key");

    const result = serializeSignatureParams(
      [component("@authority"), queryParam("Pet")],
      params,
    );
    expect(result).toBe(
      '("@authority" "@query-param";name="Pet");created=1618884473;keyid="test-key"',
    );
  });

  it("serializes bare items", () => {
    expect(serializeBareItem(42)).toBe("42");
    expect(serializeBareItem("hello")).toBe('"hello"');
    expect(serializeBareItem(true)).toBe("?1");
    expect(serializeBareItem(false)).toBe("?0");
    expect(serializeBareItem(new SFVToken("tok"))).toBe("tok");
    expect(serializeBareItem(new Uint8Array([1, 2, 3]))).toBe(":AQID:");
  });
});

describe("SFV parsing", () => {
  it("parses a simple dictionary", () => {
    const result = parseDictionary('a=1, b="hello"');
    expect(result).toHaveLength(2);
    expect(result[0].key).toBe("a");
    expect(result[0].item?.value).toBe(1);
    expect(result[1].key).toBe("b");
    expect(result[1].item?.value).toBe("hello");
  });

  it("parses dictionary with inner list", () => {
    const result = parseDictionary(
      'sig1=("@method" "@path");created=1618884473',
    );
    expect(result).toHaveLength(1);
    expect(result[0].key).toBe("sig1");
    expect(result[0].innerList).toBeDefined();
    expect(result[0].innerList!.items).toHaveLength(2);
    expect(result[0].innerList!.items[0].value).toBe("@method");
    expect(result[0].innerList!.params?.values["created"]).toBe(1618884473);
  });

  it("parses binary values", () => {
    const result = parseDictionary("sig1=:AQID:");
    expect(result[0].item?.value).toBeInstanceOf(Uint8Array);
    const bytes = result[0].item?.value as Uint8Array;
    expect(bytes).toEqual(new Uint8Array([1, 2, 3]));
  });

  it("parses boolean true member", () => {
    const result = parseDictionary("a, b=?0");
    expect(result[0].item?.value).toBe(true);
    expect(result[1].item?.value).toBe(false);
  });
});
